### PR TITLE
[5] Create cumbersome API endpoint for Sign Up 

### PR DESCRIPTION
**Pull request by junan**
## What happened

Created two API endpoints for sign up

- `/users/signup`
- `/users/verify`
 
### API 
-  `/users/signup` API will expect `password` and `email` params
- `email` param must be a valid email
- `email` param must be a uniq email
- `password` params must be more than 5 characters string
- `/users/verify` API will expect `otp` param

Just like the previously authenticated API endpoints, the `verify` API verifies the presence only of the submitted `otp` params

### Schema
- Create a migration to create `users` table
- used `comeonin` and `bcrypt_elixir` library to store hashed password
 
#### Success response
Screen Shot 2021-07-26 at 11 54 03 AM

#### Error response
Screen Shot 2021-07-30 at 10 57 58 AM
Screen Shot 2021-07-30 at 10 39 57 AM
Screen Shot 2021-07-30 at 10 39 43 AM
Screen Shot 2021-07-30 at 10 39 25 AM

### `/users/verify` API
Screen Shot 2021-07-26 at 11 57 11 AM
Screen Shot 2021-07-26 at 11 57 22 AM

### CODE REVIEW

File: lib/liveman/user/users.ex [hunk 1]
```diff
@@ -0,0 +1,46 @@
+defmodule Liveman.User.Users do
+  @email_regex_format ~r/^\S+@\S+\.\S+$/
```
Comment (junan) [inline lib/liveman/user/users.ex:2]: This simple regex I used from https://stackoverflow.com/a/201447

File: lib/liveman/user/users.ex [hunk 2]
```diff
@@ -0,0 +1,46 @@
...
+
+  def validate_registration_params(params) do
```
Comment (olivierobert) [inline lib/liveman/user/users.ex:4]: When using validation, this should be done instead at the schema level. Either use the default 	`changeset` method or create another one e.g. `registration_changeset`.

Then in the context, there can be a method called `create_user` or `register_user`:

```
def register_user(user_params) do
  User |> User.changeset(user_params) |> Repo.insert()
end
```

This will raised a changeset error when validation fails.

Then in the controller, the standard way would be:

```
case Users.register_user(params) do
  {:ok, user} ->
    json(conn, %{meta: %{detail: "An confirmation email has been sent with an OTP code"}})

  {:error, %Ecto.Changeset{} = changeset} ->
     conn
     |> put_status(:unprocessable_entity)
     |> render(ErrorView, "error.json", errors: [%{detail: changeset.errors}])
end
```

**Review by olivierobert — CHANGES_REQUESTED** [review]

File: lib/liveman_web/helpers/ecto_helpers.ex [hunk 3]
```diff
@@ -0,0 +1,34 @@
+defmodule LivemanWeb.EctoHelpers do
+  def pretty_errors(errors) do
```
Comment (junan) [inline lib/liveman_web/helpers/ecto_helpers.ex:2]: I took this code from this [article]( https://medium.com/@brain.amsterdam/prettify-ecto-errors-39512370d006)

File: lib/liveman/user/users.ex (hunk 2 above)
Comment (junan) [reply lib/liveman/user/users.ex]: Thanks @olivierobert , I changed it accordingly

File: lib/liveman/user/schemas/user_schema.ex [hunk 4]
```diff
@@ -3,15 +3,49 @@ defmodule Liveman.User.Schemas.User do
...
+    |> validate_format(:email, @email_regex_format, message: "is invalid")
+    |> validate_length(:email, max: 160)
+    |> unique_constraint(:email)
+  end
+
+  defp validate_password(changeset) do
+    changeset
+    |> validate_required([:password])
+    |> validate_length(:password, min: 6, max: 80)
+    |> update_hashed_password
```
Comment (olivierobert) [inline lib/liveman/user/schemas/user_schema.ex:37]: Minor naming preference for:

```suggestion
    |> put_hashed_password
```

It is more idiomatic as it actually puts a change into a map 💭

File: lib/liveman/user/schemas/user_schema.ex [hunk 5]
```diff
@@ -3,15 +3,49 @@ defmodule Liveman.User.Schemas.User do
 
   import Ecto.Changeset
 
+  @email_regex_format ~r/^\S+@\S+\.\S+$/
+
+  @derive {Inspect, except: [:password]}
```
Comment (olivierobert) [inline lib/liveman/user/schemas/user_schema.ex:8]: Why is this needed? I am curious as I have not used it and did not see it before 😅

File: test/liveman/user/users_test.exs [hunk 6]
```diff
@@ -0,0 +1,33 @@
+defmodule Liveman.User.UsersTest do
+  use Liveman.DataCase, async: true
+
+  alias Liveman.User.Users
+
+  describe "register_user/1" do
```
Comment (olivierobert) [inline test/liveman/user/users_test.exs:6]: The tests duplicates the schema tests. Here there could be two tests: 1) test with valid params and 2) test it returns Ecto changeset error with valid params/

**Review by olivierobert — CHANGES_REQUESTED** [review]

File: lib/liveman/user/schemas/user_schema.ex (hunk 5 above)
Comment (junan) [reply lib/liveman/user/schemas/user_schema.ex:8]: @olivierobert This one is for hide sensitive data from inspect, log etc. It will not show the `password`.  https://hexdocs.pm/elixir/Inspect.html#module-deriving
Screen Shot 2021-07-30 at 2 50 16 PM

File: test/liveman/user/users_test.exs [hunk 7]
```diff
@@ -0,0 +1,30 @@
...
+    test "creates a user successfully when the given params are valid" do
+      result = Users.register_user(%{email: "john@doe.com", password: "secret"})
```
Comment (olivierobert) [inline test/liveman/user/users_test.exs:8]: Instead of result as variable name, it could be `user`:

```suggestion
      user = Users.register_user(%{email: "john@doe.com", password: "secret"})
```

**Review by olivierobert — CHANGES_REQUESTED** [review]
Just one minor change needed

File: lib/liveman_web/router.ex [hunk 8]
```diff
@@ -19,12 +19,23 @@ defmodule LivemanWeb.Router do
...
+      end
+
+      get "/surveys", V1.SurveyController, :index
+    end
 
-    pipe_through :require_authenticated_user
-    get "/me", V1.UserController, :show
-    post "/responses", V1.AnswerController, :create
+    scope "/" do
+      pipe_through(:require_authenticated_user)
```
Comment (bterone) [inline lib/liveman_web/router.ex:34]: Does this scope still get piped through `:api`? 🤔 
I think we might need to add `:api` to this scope as well.

File: test/liveman/user/users_test.exs (hunk 6 above)
Comment (junan) [reply test/liveman/user/users_test.exs:6]: Updated, used only valid and invalid param test

**Review by bterone — APPROVED** [review]

**Review by rosle — APPROVED** [review]

**Review by olivierobert — APPROVED** [review]
